collect_app/src/main/java/org/odk/collect/android/http/CollectServerClient.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.http;
+
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import org.kxml2.io.KXmlParser;
+import org.kxml2.kdom.Document;
+import org.odk.collect.android.utilities.DocumentFetchResult;
+import org.xmlpull.v1.XmlPullParser;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashMap;
+
+import javax.inject.Inject;
+
+import timber.log.Timber;
+
+/**
+ * Common utility methods for managing the credentials associated with the
+ * request context and constructing http context, client and request with the
+ * proper parameters and OpenRosa headers.
+ *
+ * @author mitchellsundt@gmail.com
+ */
+public final class CollectServerClient {
+
+    private static final String HTTP_CONTENT_TYPE_TEXT_XML = "text/xml";
+
+    private static CollectServerClient instance;
+
+    @Inject
+    HttpInterface httpConnection;
+
+    private CollectServerClient() {
+        DaggerHttpComponent.builder().build().inject(this);
+    }
+
+    private static synchronized CollectServerClient getInstance() {
+        if (instance == null) {
+            instance = new CollectServerClient();
+        }
+        return instance;
+    }
+
+    /**
+     * Remove all credentials for accessing the specified host.
+     *
+     * @param host host to clear the credentials
+     */
+    public static void clearHostCredentials(String host) {
+        getInstance().httpConnection.clearHostCredentials(host);
+    }
+
+    /**
+     * Clears the cookie store
+     */
+    public static void clearCookieStore() {
+        getInstance().httpConnection.clearCookieStore();
+    }
+
+    /**
+     * Add credentials to a specified host
+     *
+     * @param username - The users name
+     * @param password - The password
+     * @param host - The host to add credentials to
+     */
+    public static void addCredentials(String username, String password,
+                                      String host) {
+        getInstance().httpConnection.addCredentials(username, password, host);
+    }
+
+    /**
+     * Common method for returning a parsed xml document given a url and the
+     * http context and client objects involved in the web connection.
+     */
+    public static DocumentFetchResult getXmlDocument(String urlString) {
+
+        // parse response
+        Document doc;
+
+        HttpInputStreamResult inputStreamResult;
+        try {
+            inputStreamResult = getHttpInputStream(urlString, HTTP_CONTENT_TYPE_TEXT_XML);
```

Comment:
Hey John so to make `CollectServerClient` testable once the `httpInterface` is mocked then the `getXmlDocument`method and `getHttpInputStream` could be tested right? 
Since all the other methods are just calling to the concrete `httpConnection` implementation those are verified once the `httpInterface` is tested.

Reply:
Hi, yes that's correct and I am currently writing tests for getXmlDocument and getHttpInputStream that would rely on a mocked httpInterface.

Reply:
Good stuff!